sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/AADClientRegistrationRepository.java

```diff
@@ -3,83 +3,402 @@
+    }
+
+    private AzureClientRegistration emptyAzureClientRegistration() {
+        ClientRegistration.Builder clientRegistrationBuilder = toClientRegistrationBuilder(
+            AZURE_CLIENT_REGISTRATION_ID, defaultAzureAuthorizationClientProperties());
+        return new AzureClientRegistration(clientRegistrationBuilder.build(), new HashSet<>());
+    }
+
+    private AuthorizationClientProperties defaultAzureAuthorizationClientProperties() {
+        AuthorizationClientProperties result = new AuthorizationClientProperties();
+        result.setAuthorizationGrantType(AUTHORIZATION_CODE);
+        return result;
+    }
 
-        allClients = new HashMap<>();
-        addClientRegistration(azureClient.getClient());
-        for (ClientRegistration c : otherClients) {
-            addClientRegistration(c);
+    private String toClientId(AADAuthenticationProperties properties) {
+        return Optional.of(properties)
+                       .map(AADAuthenticationProperties::getClientId)
+                       .filter(StringUtils::hasText)
+                       .orElse("client-id-not-configured");
+    }
+
+    private List<String> toScopes(AuthorizationClientProperties clientProperties) {
+        List<String> result = clientProperties.getScopes();
+        if (clientProperties.isOnDemand()) {
+            if (!result.contains("openid")) {
+                result.add("openid");
+            }
+            if (!result.contains("profile")) {
+                result.add("profile");
+            }
         }
+        return result;
+    }
+
+    private Map<String, Object> providerConfigurationMetadata(AADAuthorizationServerEndpoints endpoints) {
+        Map<String, Object> result = new LinkedHashMap<>();
+        String endSessionEndpoint = endpoints.endSessionEndpoint();
+        result.put("end_session_endpoint", endSessionEndpoint);
+        return result;
     }
 
-    private void addClientRegistration(ClientRegistration client) {
-        allClients.put(client.getRegistrationId(), client);
+    public static int resourceServerCount(Set<String> scopes) {
+        return (int) scopes.stream()
+                           .filter(scope -> scope.contains("/"))
+                           .map(scope -> scope.substring(0, scope.lastIndexOf('/')))
+                           .distinct()
+                           .count();
+    }
+
+    private Set<String> authorizationCodeScopes() {
+        Set<String> result = accessTokenScopes();
+        for (AuthorizationClientProperties authProperties : properties.getAuthorizationClients().values()) {
+            if (!authProperties.isOnDemand()
+                && isDefaultAuthorizationGrantType(authProperties)) {
+                result.addAll(authProperties.getScopes());
+            }
+        }
+        return result;
+    }
+
+    private boolean isDefaultAuthorizationGrantType(AuthorizationClientProperties authProperties) {
+        return authProperties.getAuthorizationGrantType() == null
```

Comment:
`authProperties.getAuthorizationGrantType()` should never be null now.

Reply:
In the new scenario, it should not be null, but it still can be null for the Web application only scenario.